### DISCUSSION TITLE
How to deal with multiple equal IPs?

### DISCUSSION THREAD

**Discussion by ZSasha0** [netbox-community/netbox#5731] [Q&A]
Hi guys,
Can anyone give me a hint about how you deal with situation when the same IP exists on a multiple devices, please?

Generally speaking, an IP should be unique (per VRF or global).
But sometimes it happens that more than one device may have the same IPs configured - for example when a company is migrating from an old router to a new one. 
In that case the old interface may be just shut but its whole config may be left intact. 
At the same time a new device is installed and duplicated config (with duplicated IPs) is applied.

I hoped that making the "old" IP's status as "Reserved" or even "Deprecated"  would let me create another "Active" IP but that did not work.
I am aware about two "workarounds" - A) make it "Anycast" or B) disabling IP uniqueness check in config would let me do that but it comes with significant "side effects" IMO.

Wondering how you tackle situations like that?

Cheers.

**Comment by candlerb** [comment]
You're right: making it "Anycast" (or "VIP" etc) will do the trick.  Another option is to create a fake VRF, e.g. "migrated" or "shutdown", and put the old IP address in that VRF.

Personally, I simply reassign the IP address from the old device interface directly to the new device interface, instead of creating a new one. This does lose the record that the address *was* assigned to the old device (except in ChangeLog); however I see Netbox more as a record of how the network is *now* rather than as a historical record.  If an interface is intentionally shut down, I don't consider the address assigned there, even if there's stale config.

In your use case, you are keeping the old address assigned (albeit with the interface shut down).  So really you *are* assigning duplicate IP addresses.  As you say, you could disable the uniqueness check, but that will allow lots of other errors that you don't want to allow.

**Reply by ZSasha0** [reply]
Do you think we can ask for this feature to be implemented - allowing multiple same IPs as long as they are "disabled" and/or "reserved"?
If NetBox supposed to be a "Source of Truth" as stated officially, my config example is perfectly valid and it goes well with a real world - real devices do let us put multiple same IPs as long as the interface remains shut, so why not NetBox?

**Reply by candlerb** [reply]
In principle, anything is possible - but it may not be worth the effort, or even desirable.  In particular, if I set an address to "reserved", then I don't want to allow anyone else to be permitted to use this IP address, *by definition*.

The definition of "deprecated" could be changed to mean "someone else can use this IP now".  This would require some logic where a duplicate is allowed as long as one address is "deprecated" and one is "non-deprecated", or vice versa, and you'd have to deal correctly with all the possible state transitions in between.  This would then involve two fields, the "role" and "status", and you'd have to consider all the combinations.

I made a quick check and found that even Netbox's simple duplicate IP address on "role" is currently broken - see #5735.  And that's without considering that it's not enforced at the database layer, so in principle there's a race if two users commit at the same time.

If you wanted to allow duplicates dependent on the status of the *interface* to which the address was attached then the logic would be far more complex, and again not necessarily desirable.  For example, if I shutdown a device or interface temporarily, that doesn't necessarily mean I want some other device to be assigned its IP address in the mean time.

**Reply by DanSheps** [reply]
Right now, there is no database enforced uniqueness, so it probably would be fairly trivial to sort this out in the code.  However we would need a proper FR to consider it and as you can see we are backed up on FRs right now.
